README.md
# CITVYJS (citvyjs)

Pure JavaScript Bitshares library for node.js and browsers. Can be used to construct, sign and broadcast transactions in JavaScript, and to easily obtain data from the blockchain via public apis.

Most of this code was written by [jcalfee](https://github.com/jcalfee), my work was mostly just repackaging to a discrete npm package.

Security issues might be eligible for a bounty through the [HackTheDex](https://hackthedex.io) program.

## Setup

This library can be obtained through npm:

```
npm install citvyjs
```

## Usage

Three sub-libraries are included: `ECC`, `Chain` and `Serializer`. Generally only the `ECC` and `Chain` libraries need to be used directly.

### Chain

This library provides utility functions to handle blockchain state as well as a login class that can be used for simple login functionality using a specific key seed.

#### Login

The login class uses the following format for keys:

```
keySeed = accountName + role + password
```

Using this seed, private keys are generated for either the default roles `active, owner, memo`, or as specified. A minimum password length of 12 characters is enforced, but an even longer password is recommended. Three methods are provided:

```
generateKeys(account, password, [roles])
checkKeys(account, password, auths)
signTransaction(tr)
```

The auths object should contain the auth arrays from the account object. An example is this:

```
{
    active: [
        ["GPH5Abm5dCdy3hJ1C5ckXkqUH2Me7dXqi9Y7yjn9ACaiSJ9h8r8mL", 1]
    ]
}
```

If checkKeys is successful, you can use signTransaction to sign a TransactionBuilder transaction using the private keys for that account.

#### State container

The Chain library contains a complete state container called the ChainStore. The ChainStore will automatically configure the `set_subscribe_callback` and handle any incoming state changes appropriately. It uses Immutable.js for storing the state, so all objects are return as immutable objects. It has its own `subscribe` method that can be used to register a callback that will be called whenever a state change happens.

The ChainStore has several useful methods to retrieve, among other things, objects, assets and accounts using either object ids or asset/account names. These methods are synchronous and will return `undefined` to indicate fetching in progress, and `null` to indicate that the object does not exist.

```
import {Apis} from "citvyjs-ws";
var {ChainStore} = require("citvyjs");

Apis.instance("wss://eu.nodes.bitshares.ws", true).init_promise.then((res) => {
    console.log("connected to:", res[0].network);
    ChainStore.init().then(() => {
        ChainStore.subscribe(updateState);
    });
});

let dynamicGlobal = null;
function updateState(object) {
    dynamicGlobal = ChainStore.getObject("2.1.0");
    console.log("ChainStore object update\n", dynamicGlobal ? dynamicGlobal.toJS() : dynamicGlobal);
}
```

### ECC

The ECC library contains all the crypto functions for private and public keys as well as transaction creation/signing.

#### Private keys

As a quick example, here's how to generate a new private key from a seed (a brainkey for example):

```
var {PrivateKey, key} = require("citvyjs");

let seed = "THIS IS A TERRIBLE BRAINKEY SEED WORD SEQUENCE";
let pkey = PrivateKey.fromSeed( key.normalize_brainKey(seed) );

console.log("\nPrivate key:", pkey.toWif());
console.log("Public key :", pkey.toPublicKey().toString(), "\n");
```

#### Transactions

TODO transaction signing example

## ESDoc (beta)

```bash
npm i -g esdoc esdoc-es7-plugin
esdoc -c ./esdoc.json
open out/esdoc/index.html
```

## Binaries / Browserified bundles

Please have a [look here](https://github.com/bitshares/citvyjs/releases) to find your desired release.
If you want to build the binaries yourself you can clone this repository and run `npm install`. It will
create

-   Browserified version `build/citvyjs.js`
-   Browserified and minified (babel) version `build/citvyjs.min.js`
-   CommonJS version using wepback `build/citvyjs.cjs`


CHANGELOG.md
## v0.4.0
- Add AccountLogin class
- Clean up imports
- Use babel for tests
## v0.3.0
- Extract websocket library to separate citvyjs-ws lib


lib/ecc/src/address.js
import assert from "assert";
import {ChainConfig} from "citvyjs-ws";
import {sha256, sha512, ripemd160} from "./hash";
import {encode, decode} from "bs58";
import deepEqual from "deep-equal";
const Buffer = require("safe-buffer").Buffer;

/** Addresses are shortened non-reversable hashes of a public key.  The full PublicKey is preferred.
 */
class Address {
    constructor(addy) {
        this.addy = addy;
    }

    static fromBuffer(buffer) {
        var _hash = sha512(buffer);
        var addy = ripemd160(_hash);
        return new Address(addy);
    }

    static fromString(string, address_prefix = ChainConfig.address_prefix) {
        var prefix = string.slice(0, address_prefix.length);
        assert.equal(
            address_prefix,
            prefix,
            `Expecting key to begin with ${address_prefix}, instead got ${prefix}`
        );
        var addy = string.slice(address_prefix.length);
        addy = new Buffer(decode(addy), "binary");
        var checksum = addy.slice(-4);
        addy = addy.slice(0, -4);
        var new_checksum = ripemd160(addy);
        new_checksum = new_checksum.slice(0, 4);
        var isEqual = deepEqual(checksum, new_checksum); //, 'Invalid checksum'
        if (!isEqual) {
            throw new Error("Checksum did not match");
        }
        return new Address(addy);
    }

    /** @return Address - Compressed PTS format (by default) */
    static fromPublic(public_key, compressed = true, version = 56) {
        var sha2 = sha256(public_key.toBuffer(compressed));
        var rep = ripemd160(sha2);
        var versionBuffer = Buffer.alloc(1);
        versionBuffer.writeUInt8(0xff & version, 0);
        var addr = Buffer.concat([versionBuffer, rep]);
        var check = sha256(addr);
        check = sha256(check);
        var buffer = Buffer.concat([addr, check.slice(0, 4)]);
        return new Address(ripemd160(buffer));
    }

    toBuffer() {
        return this.addy;
    }

    toString(address_prefix = ChainConfig.address_prefix) {
        var checksum = ripemd160(this.addy);
        var addy = Buffer.concat([this.addy, checksum.slice(0, 4)]);
        return address_prefix + encode(addy);
    }
}

export default Address;


lib/ecc/src/PublicKey.js
import BigInteger from "bigi";
import {Point, getCurveByName} from "ecurve";
const secp256k1 = getCurveByName("secp256k1");
import {encode, decode} from "bs58";
import {sha256, sha512, ripemd160} from "./hash";
import {ChainConfig} from "citvyjs-ws";
import assert from "assert";
import deepEqual from "deep-equal";
const Buffer = require("safe-buffer").Buffer;
const {G, n} = secp256k1;

class PublicKey {
    /** @param {Point} public key */
    constructor(Q) {
        this.Q = Q;
    }

    static fromBinary(bin) {
        return PublicKey.fromBuffer(Buffer.from(bin, "binary"));
    }

    static fromBuffer(buffer) {
        if (
            buffer.toString("hex") ===
            "000000000000000000000000000000000000000000000000000000000000000000"
        )
            return new PublicKey(null);
        return new PublicKey(Point.decodeFrom(secp256k1, buffer));
    }

    toBuffer(compressed = this.Q ? this.Q.compressed : null) {
        if (this.Q === null)
            return Buffer.from(
                "000000000000000000000000000000000000000000000000000000000000000000",
                "hex"
            );
        return this.Q.getEncoded(compressed);
    }

    static fromPoint(point) {
        return new PublicKey(point);
    }

    toUncompressed() {
        var buf = this.Q.getEncoded(false);
        var point = Point.decodeFrom(secp256k1, buf);
        return PublicKey.fromPoint(point);
    }

    /** bts::blockchain::address (unique but not a full public key) */
    toBlockchainAddress() {
        var pub_buf = this.toBuffer();
        var pub_sha = sha512(pub_buf);
        return ripemd160(pub_sha);
    }

    /** Alias for {@link toPublicKeyString} */
    toString(address_prefix = ChainConfig.address_prefix) {
        return this.toPublicKeyString(address_prefix);
    }

    /**
        Full public key
        {return} string
    */
    toPublicKeyString(address_prefix = ChainConfig.address_prefix) {
        var pub_buf = this.toBuffer();
        var checksum = ripemd160(pub_buf);
        var addy = Buffer.concat([pub_buf, checksum.slice(0, 4)]);
        return address_prefix + encode(addy);
    }

    /**
        @arg {string} public_key - like GPHXyz...
        @arg {string} address_prefix - like GPH
        @return PublicKey or `null` (if the public_key string is invalid)
    */
    static fromPublicKeyString(
        public_key,
        address_prefix = ChainConfig.address_prefix
    ) {
        try {
            return PublicKey.fromStringOrThrow(public_key, address_prefix);
        } catch (e) {
            return null;
        }
    }

    /**
        @arg {string} public_key - like GPHXyz...
        @arg {string} address_prefix - like GPH
        @throws {Error} if public key is invalid
        @return PublicKey
    */
    static fromStringOrThrow(
        public_key,
        address_prefix = ChainConfig.address_prefix
    ) {
        if (public_key.Q === null)
            public_key =
                address_prefix + "1111111111111111111111111111111114T1Anm"; // null key
        var prefix = public_key.slice(0, address_prefix.length);
        assert.equal(
            address_prefix,
            prefix,
            `Expecting key to begin with ${address_prefix}, instead got ${prefix}`
        );
        public_key = public_key.slice(address_prefix.length);

        public_key = Buffer.from(decode(public_key), "binary");
        var checksum = public_key.slice(-4);
        public_key = public_key.slice(0, -4);
        var new_checksum = ripemd160(public_key);
        new_checksum = new_checksum.slice(0, 4);
        var isEqual = deepEqual(checksum, new_checksum); //, 'Invalid checksum'
        if (!isEqual) {
            throw new Error("Checksum did not match");
        }
        return PublicKey.fromBuffer(public_key);
    }

    toAddressString(address_prefix = ChainConfig.address_prefix) {
        var pub_buf = this.toBuffer();
        var pub_sha = sha512(pub_buf);
        var addy = ripemd160(pub_sha);
        var checksum = ripemd160(addy);
        addy = Buffer.concat([addy, checksum.slice(0, 4)]);
        return address_prefix + encode(addy);
    }

    toPtsAddy() {
        var pub_buf = this.toBuffer();
        var pub_sha = sha256(pub_buf);
        var addy = ripemd160(pub_sha);
        addy = Buffer.concat([Buffer.from([0x38]), addy]); //version 56(decimal)

        var checksum = sha256(addy);
        checksum = sha256(checksum);

        addy = Buffer.concat([addy, checksum.slice(0, 4)]);
        return encode(addy);
    }

    child(offset) {
        assert(Buffer.isBuffer(offset), "Buffer required: offset");
        assert.equal(offset.length, 32, "offset length");

        offset = Buffer.concat([this.toBuffer(), offset]);
        offset = sha256(offset);

        let c = BigInteger.fromBuffer(offset);

        if (c.compareTo(n) >= 0)
            throw new Error("Child offset went out of bounds, try again");

        let cG = G.multiply(c);
        let Qprime = this.Q.add(cG);

        if (secp256k1.isInfinity(Qprime))
            throw new Error(
                "Child offset derived to an invalid key, try again"
            );

        return PublicKey.fromPoint(Qprime);
    }

    /* <HEX> */

    toByteBuffer() {
        var b = new ByteBuffer(
            ByteBuffer.DEFAULT_CAPACITY,
            ByteBuffer.LITTLE_ENDIAN
        );
        this.appendByteBuffer(b);
        return b.copy(0, b.offset);
    }

    static fromHex(hex) {
        return PublicKey.fromBuffer(Buffer.from(hex, "hex"));
    }

    toHex() {
        return this.toBuffer().toString("hex");
    }

    static fromPublicKeyStringHex(hex) {
        return PublicKey.fromPublicKeyString(Buffer.from(hex, "hex"));
    }

    /* </HEX> */
}

export default PublicKey;


lib/chain/src/TransactionHelper.js
var helper = {};

import secureRandom from "secure-random";

import {Long} from "bytebuffer";

import {Signature} from "../../ecc";
import {ops} from "../../serializer";
import {Apis} from "citvyjs-ws";

helper.unique_nonce_entropy = null;
helper.unique_nonce_uint64 = function() {
    var entropy = (helper.unique_nonce_entropy = (() => {
        if (helper.unique_nonce_entropy === null) {
            //console.log('... secureRandom.randomUint8Array(1)[0]',secureRandom.randomUint8Array(1)[0])
            return parseInt(secureRandom.randomUint8Array(1)[0]);
        } else {
            return ++helper.unique_nonce_entropy % 256;
        }
    })());
    var long = Long.fromNumber(Date.now());
    //console.log('unique_nonce_uint64 date\t',ByteBuffer.allocate(8).writeUint64(long).toHex(0))
    //console.log('unique_nonce_uint64 entropy\t',ByteBuffer.allocate(8).writeUint64(Long.fromNumber(entropy)).toHex(0))
    long = long.shiftLeft(8).or(Long.fromNumber(entropy));
    //console.log('unique_nonce_uint64 shift8\t',ByteBuffer.allocate(8).writeUint64(long).toHex(0))
    return long.toString();
};

/* Todo, set fees */
helper.to_json = function(tr, broadcast = false) {
    return (function(tr, broadcast) {
        var tr_object = ops.signed_transaction.toObject(tr);
        if (broadcast) {
            var net = Apis.instance().network_api();
            console.log("... tr_object", JSON.stringify(tr_object));
            return net.exec("broadcast_transaction", [tr_object]);
        } else {
            return tr_object;
        }
    })(tr, broadcast);
};

helper.signed_tr_json = function(tr, private_keys) {
    var tr_buffer = ops.transaction.toBuffer(tr);
    tr = ops.transaction.toObject(tr);
    tr.signatures = (() => {
        var result = [];
        for (
            var i = 0;
            0 < private_keys.length
                ? i < private_keys.length
                : i > private_keys.length;
            0 < private_keys.length ? i++ : i++
        ) {
            var private_key = private_keys[i];
            result.push(Signature.signBuffer(tr_buffer, private_key).toHex());
        }
        return result;
    })();
    return tr;
};

helper.expire_in_min = function(min) {
    return Math.round(Date.now() / 1000) + min * 60;
};

helper.seconds_from_now = function(timeout_sec) {
    return Math.round(Date.now() / 1000) + timeout_sec;
};

/**
    Print to the console a JSON representation of any object in
    @graphene/serializer { types }
*/
helper.template = function(
    serializer_operation_type_name,
    debug = {use_default: true, annotate: true}
) {
    var so = ops[serializer_operation_type_name];
    if (!so) {
        throw new Error(
            `unknown serializer_operation_type ${serializer_operation_type_name}`
        );
    }
    return so.toObject(undefined, debug);
};

helper.new_operation = function(serializer_operation_type_name) {
    var so = ops[serializer_operation_type_name];
    if (!so) {
        throw new Error(
            `unknown serializer_operation_type ${serializer_operation_type_name}`
        );
    }
    var object = so.toObject(undefined, {use_default: true, annotate: true});
    return so.fromObject(object);
};

helper.instance = function(ObjectId) {
    return ObjectId.substring("0.0.".length);
};

export default helper;


lib/ecc/src/KeyUtils.js
import PrivateKey from "./PrivateKey";
import PublicKey from "./PublicKey";
import Address from "./address";
import Aes from "./aes";
import {sha256, sha512} from "./hash";
// import dictionary from './dictionary_en';
import secureRandom from "secure-random";
import {ChainConfig} from "citvyjs-ws";
const Buffer = require("safe-buffer").Buffer;

// hash for .25 second
var HASH_POWER_MILLS = 250;

const key = {
    /** Uses 1 second of hashing power to create a key/password checksum.  An
    implementation can re-call this method with the same password to re-match
    the strength of the CPU (either after moving from a desktop to a mobile,
    mobile to desktop, or N years from now when CPUs are presumably stronger).

    A salt is used for all the normal reasons...

    @return object {
        aes_private: Aes,
        checksum: "{hash_iteration_count},{salt},{checksum}"
    }
    */
    aes_checksum(password) {
        if (!(typeof password === "string")) {
            throw new "password string required"();
        }
        var salt = secureRandom.randomBuffer(4).toString("hex");
        var iterations = 0;
        var secret = salt + password;
        // hash for .1 second
        var start_t = Date.now();
        while (Date.now() - start_t < HASH_POWER_MILLS) {
            secret = sha256(secret);
            iterations += 1;
        }

        var checksum = sha256(secret);
        var checksum_string = [
            iterations,
            salt.toString("hex"),
            checksum.slice(0, 4).toString("hex")
        ].join(",");

        return {
            aes_private: Aes.fromSeed(secret),
            checksum: checksum_string
        };
    },

    /** Provide a matching password and key_checksum.  A "wrong password"
    error is thrown if the password does not match.  If this method takes
    much more or less than 1 second to return, one should consider updating
    all encyrpted fields using a new key.key_checksum.
    */
    aes_private(password, key_checksum) {
        var [iterations, salt, checksum] = key_checksum.split(",");
        var secret = salt + password;
        for (
            var i = 0;
            0 < iterations ? i < iterations : i > iterations;
            0 < iterations ? i++ : i++
        ) {
            secret = sha256(secret);
        }
        var new_checksum = sha256(secret);
        if (!(new_checksum.slice(0, 4).toString("hex") === checksum)) {
            throw new Error("wrong password");
        }
        return Aes.fromSeed(secret);
    },

    /**
        A week random number generator can run out of entropy.  This should ensure even the worst random number implementation will be reasonably safe.

        @param1 string entropy of at least 32 bytes
    */
    random32ByteBuffer(entropy = this.browserEntropy()) {
        if (!(typeof entropy === "string")) {
            throw new Error("string required for entropy");
        }

        if (entropy.length < 32) {
            throw new Error("expecting at least 32 bytes of entropy");
        }

        var start_t = Date.now();

        while (Date.now() - start_t < HASH_POWER_MILLS)
            entropy = sha256(entropy);

        var hash_array = [];
        hash_array.push(entropy);

        // Hashing for 1 second may helps the computer is not low on entropy (this method may be called back-to-back).
        hash_array.push(secureRandom.randomBuffer(32));

        return sha256(Buffer.concat(hash_array));
    },

    suggest_brain_key: function(
        dictionary = ",",
        entropy = this.browserEntropy()
    ) {
        var randomBuffer = this.random32ByteBuffer(entropy);

        var word_count = 16;
        var dictionary_lines = dictionary.split(",");

        if (!(dictionary_lines.length === 49744)) {
            throw new Error(
                `expecting ${49744} but got ${
                    dictionary_lines.length
                } dictionary words`
            );
        }

        var brainkey = [];
        var end = word_count * 2;

        for (let i = 0; i < end; i += 2) {
            // randomBuffer has 256 bits / 16 bits per word == 16 words
            var num = (randomBuffer[i] << 8) + randomBuffer[i + 1];

            // convert into a number between 0 and 1 (inclusive)
            var rndMultiplier = num / Math.pow(2, 16);
            var wordIndex = Math.round(dictionary_lines.length * rndMultiplier);

            brainkey.push(dictionary_lines[wordIndex]);
        }
        return this.normalize_brainKey(brainkey.join(" "));
    },

    get_random_key(entropy) {
        return PrivateKey.fromBuffer(this.random32ByteBuffer(entropy));
    },

    get_brainPrivateKey(brainKey, sequence = 0) {
        if (sequence < 0) {
            throw new Error("invalid sequence");
        }
        if (brainKey.trim() === "") {
            throw new Error("empty brain key");
        }
        brainKey = key.normalize_brainKey(brainKey);
        return PrivateKey.fromBuffer(sha256(sha512(brainKey + " " + sequence)));
    },

    // Turn invisible space like characters into a single space
    normalize_brainKey(brainKey) {
        if (!(typeof brainKey === "string")) {
            throw new Error("string required for brainKey");
        }

        brainKey = brainKey.trim();
        if (brainKey === "") {
            throw new Error("empty brain key");
        }
        return brainKey.split(/[\t\n\v\f\r ]+/).join(" ");
    },

    browserEntropy() {
        var entropyStr = "";
        try {
            entropyStr =
                new Date().toString() +
                " " +
                window.screen.height +
                " " +
                window.screen.width +
                " " +
                window.screen.colorDepth +
                " " +
                " " +
                window.screen.availHeight +
                " " +
                window.screen.availWidth +
                " " +
                window.screen.pixelDepth +
                navigator.language +
                " " +
                window.location +
                " " +
                window.history.length;

            for (var i = 0, mimeType; i < navigator.mimeTypes.length; i++) {
                mimeType = navigator.mimeTypes[i];
                entropyStr +=
                    mimeType.description +
                    " " +
                    mimeType.type +
                    " " +
                    mimeType.suffixes +
                    " ";
            }
            console.log("INFO\tbrowserEntropy gathered");
        } catch (error) {
            //nodejs:ReferenceError: window is not defined
            entropyStr = sha256(new Date().toString());
        }

        var b = Buffer.from(entropyStr);
        entropyStr += b.toString("binary") + " " + new Date().toString();
        return entropyStr;
    },

    // @return array of 5 legacy addresses for a pubkey string parameter.
    addresses(pubkey, address_prefix = ChainConfig.address_prefix) {
        var public_key = PublicKey.fromPublicKeyString(pubkey, address_prefix);
        // S L O W
        var address_string = [
            Address.fromPublic(public_key, false, 0).toString(address_prefix), // btc_uncompressed
            Address.fromPublic(public_key, true, 0).toString(address_prefix), // btc_compressed
            Address.fromPublic(public_key, false, 56).toString(address_prefix), // pts_uncompressed
            Address.fromPublic(public_key, true, 56).toString(address_prefix), // pts_compressed
            public_key.toAddressString(address_prefix) // bts_short, most recent format
        ];
        return address_string;
    }
};

export default key;
